### ISSUE TITLE
Custom UTC DateTime Type

### ISSUE THREAD

**Issue by sf-wilbur** [pydantic/pydantic#9127]
### Discussed in https://github.com/pydantic/pydantic/discussions/9126

<div type='discussions-op-text'>

<sup>Originally posted by **sf-wilbur** March 27, 2024</sup>
I am trying to create a script that will allow me to use UTCDateTime when processing geomagnetic data as I was doing so with pydantic version 1 and the json_encoders function. I updated to Pydantic version 2 and changed the necessary functions and method names thorughout my whole program but am unable to use the new tools, such as annotated or generate_core_schema to allow the Pydantic Basemodel to accept UTCDatetimes. I need to be able to provide UTCDatetime to multiple Pydantic basemodels throughout my program and would prefer to not have to iteratively go through my whole program and change each instance of UTCDatetime to a new defined custom type. 

Here is what I have tried but this only generates an unable to generate core_scheme for obspy.utcdatatime.utcdatetime error. 

`from datetime import datetime
from typing import Any, List
from obspy.core.utcdatetime import UTCDateTime
from pydantic import BaseModel, GetCoreSchemaHandler
from pydantic_core import CoreSchema, core_schema

class CustomDate(BaseModel):
    """Custom date handling for strings and ObsPy's UTCDateTime."""
    allowed_formats: List[str] = ['%Y-%m-%d', '%d-%m-%Y', '%m/%d/%Y']

    class config:
        arbitrary_types_allowed = True 
        json_encoders = {
        UTCDateTime: lambda v: v.isoformat()  # Serializing UTCDateTime to ISO format string
    }

    @classmethod
    def try_parse_date(cls, v: Any, info: core_schema.ValidationInfo) -> Any:
        if isinstance(v, str):
            for fmt in cls.allowed_formats:
                try:
                    return cls(datetime.strptime(v, fmt).date())
                except ValueError:
                    continue
            raise ValueError("Unable to parse date from string using allowed formats.")
        
        if isinstance(v, UTCDateTime):
            return cls(v.datetime.date())
        else:
            raise ValueError("Unsupported type for date conversion.")

    @classmethod
    def __get_pydantic_core_schema__(
        cls, source_type: Any, handler: GetCoreSchemaHandler) ->CoreSchema:
        
        # assert source_type is UTCDateTime
        return core_schema.with_info_after_validator_function(
        cls.model_validate, handler.generate_schema(UTCDateTime),
        core_schema.str_schema(), 
        serialization=core_schema.to_string_ser_schema()
        )
   
# Example usage
if __name__ == "__main__":
    custom_date = CustomDate.try_parse_date("2023-03-21")
    print(custom_date)

`

Here is my original code that was working before updating to version 2 I added the version 1 handles to the Pydantic tools as an interim fix but that isn't working either. 

`"""Configure pydantic to allow UTCDateTime attributes on models.
"""

from datetime import datetime
from typing import Any, Callable, Dict, List, Tuple, TypeVar, Union

# FIXME: migrating to pydantic v2 should make this unnecessary
#        (other changes will be needed though)
from fastapi.encoders import ENCODERS_BY_TYPE

from obspy import UTCDateTime
from pydantic.v1.errors import PydanticValueError
import pydantic.v1.json
import pydantic.v1.schema
import pydantic.v1.validators

# placeholder type for register_custom_pydantic_type method
CustomType = TypeVar("CustomType")

def register_custom_pydantic_type(
    custom_type: CustomType,
    encoder: Callable[[CustomType], Any],
    json_schema: Dict,
    parsers: List[Callable[[Any], CustomType]],
):
    try:
        if custom_type.__custom_pydantic_type__:
            # already registered
            return
    except AttributeError:
        # not registered yet
        pass
    # add encoder
    pydantic.v1.json.ENCODERS_BY_TYPE[custom_type] = encoder

    # FIXME: migrating to pydantic v2 should make this unnecessary
    #        (other changes will be needed though)
    ENCODERS_BY_TYPE[custom_type] = encoder

    # add openapi mapping
    pydantic.v1.schema.field_class_to_schema += ((custom_type, json_schema),)
    # add validator
    pydantic.v1.validators._VALIDATORS.append((custom_type, parsers))
    # mark as installed
    custom_type.__custom_pydantic_type__ = True

class UTCDateTimeError(PydanticValueError):
    msg_template = "invalid date-time format"

def format_utcdatetime(o: UTCDateTime) -> str:
    return o.strftime("%Y-%m-%dT%H:%M:%S.%fZ")

def parse_utcdatetime(
    value: Union[datetime, float, int, str, UTCDateTime]
) -> UTCDateTime:
    try:
        return UTCDateTime(value)
    except:
        raise UTCDateTimeError()

register_custom_pydantic_type(
    UTCDateTime,
    encoder=format_utcdatetime,
    json_schema={"type": "string", "format": "date-time"},
    parsers=[parse_utcdatetime],
)
`</div>

**Comment by sydney-runkle** [comment]
Hi @sf-wilbur,

Could you please reformat your question such that the intended code is in blocks? A bit hard to follow at the moment.

**Comment by sydney-runkle** [comment]
Hi @sf-wilbur,

Thanks for your question! You're on the right track with implementing `__get_pydantic_core_schema__`. I'll note, `json_encoders` is deprecated, so I'd advise against using that resource.

An even more simple approach might be to create an annotated type `MyCustomDateTime` that uses a `PlainValidator` and `PlainSerializer` with any custom logic you have. Marking as closed for now (this is more of a question / discussion), but happy to continue to iterate and help you out with your issue :).

**Comment by sydney-runkle** [comment]
@sf-wilbur,

No problem at all! I'll take a closer look later this week and come up with a snippet for you :).

**Comment by sf-wilbur** [comment]
Hi @sydney-runkle 

I managed to get this far so far. It still doesn't work when I try to implement it for the rest of my program. I it possible to use this code so that any time a Basemodel is instantiated it will recognize if a variable is assigned as UTCDateTime? 

```
from typing import Any

from pydantic_core import core_schema
from typing_extensions import Annotated
from typing import Dict, List, Union

from pydantic.json_schema import JsonSchemaValue
from datetime import datetime, date, timezone
from obspy import UTCDateTime
from typing_extensions import Annotated
from pydantic import BaseModel, ConfigDict, Field, ValidationError, field_validator, TypeAdapter, AfterValidator, validator

# # Define a custom type for UTCDateTime
# class UTCDateTime:
#     def __init__(self, value):
#         # Implement your logic to validate and handle UTC date and time
#         # For example, you can use datetime.datetime.strptime to parse the input string
#         self.value = value

# Define your custom type for UTCDateTime
class UTCDateTime(Annotated[UTCDateTime, Field(description="UTC timestamp")]):
    
    @classmethod
    def __get_validators__(cls):
        yield cls.validate


    @classmethod
    def validate(cls, value, self):
        if isinstance(value, datetime):
            return value.astimezone(timezone.utc)
        elif isinstance(value, str):
            try:
                return datetime.fromisoformat(value).astimezone(timezone.utc)
            except ValueError:
                raise ValueError("Invalid UTC datetime string format")
        else:
            raise ValueError("UTCDateTime must be a datetime object or a valid datetime string")

        

# Create a TypeAdapter for UTCDateTime
ta = TypeAdapter(UTCDateTime)

```

**Comment by sydney-runkle** [comment]
Hi @sf-wilbur,

I think this example would come in handy here :)

https://docs.pydantic.dev/dev/examples/validators/#using-custom-validators-with-annotated-metadata
